pkg/services/auth/auth.go

```diff
@@ -0,0 +1,250 @@
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/fluxcd/go-git-providers/gitprovider"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/weaveworks/weave-gitops/pkg/flux"
+	"github.com/weaveworks/weave-gitops/pkg/git"
+	"github.com/weaveworks/weave-gitops/pkg/gitproviders"
+	"github.com/weaveworks/weave-gitops/pkg/logger"
+	"github.com/weaveworks/weave-gitops/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// BlockingCLIAuthHandler takes over the terminal experience and returns a token when the user completes the flow.
+type BlockingCLIAuthHandler func(context.Context, io.Writer) (string, error)
+
+func NewAuthCLIHandler(name gitproviders.GitProviderName) (BlockingCLIAuthHandler, error) {
+	switch name {
+	case gitproviders.GitProviderGitHub:
+		return NewGithubDeviceFlowHandler(http.DefaultClient), nil
+
+	}
+
+	return nil, fmt.Errorf("unsupported auth provider \"%s\"", name)
+}
+
+type AuthService struct {
+	logger     logger.Logger
+	fluxClient flux.Flux
+	// Note that this is a k8s go-client, NOT a wego kube.Kube interface.
+	// That interface wasn't providing any valuable abstraction for this service.
+	k8sClient   client.Client
+	gitProvider gitproviders.GitProvider
+	gitClient   git.Git
+}
+
+// NewAuthService constructs an auth service for doing git and GitProvider operations with authenticated clients.
+func NewAuthService(fluxClient flux.Flux, k8sClient client.Client, providerName gitproviders.GitProviderName, l logger.Logger, token string) (*AuthService, error) {
+	provider, err := gitproviders.New(gitproviders.Config{Provider: providerName, Token: token})
+	if err != nil {
+		return nil, fmt.Errorf("error creating git provider client: %w", err)
+	}
+	return &AuthService{
+		logger:      l,
+		fluxClient:  fluxClient,
+		k8sClient:   k8sClient,
+		gitProvider: provider,
+		// We do not have a gitClient yet. We need to get or create deploy keys first.
+		// This will get populated by the SetupDeployKeys call.
+		// Listing it here for the sake of readability (it would still be nil without this entry).
+		gitClient: nil,
+	}, nil
+}
+
+// SetupDeployKey instruments the AuthService with the neccesary data to provide a git client that uses deploy keys.
+// The user MUST call this function before calling .GitClient(), else an error will be returned.
```

Comment:
Does it make sense to just return the gitClient from here instead of having a separate "getter" for the git client?  It would simplify the consumption of this implementation.  Looking at whee we use `SetupDeployKey` it is always called before `GitClient`.
The following code can be simplified from:
```
if err := authsvc.SetupDeployKey(ctx, name, targetName, normalizedUrl); err != nil {
    return fmt.Errorf("error setting up deploy keys: %w", err)
}

gitClient, err = authsvc.GitClient()
if err != nil {
    return fmt.Errorf("error getting authenticated git client: %w", err)
}
```
to:
```
gitClient, err = authsvc.SetupGitClient(ctx, name, targetName, normalizedUrl)
if err != nil {
    return fmt.Errorf("error setting up git client: %w", err)
}
```

Reply:
This could also eliminate the need for `ErrNoDeployKeysSetup`

Reply:
Refactored to return the `.GitClient()`